src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,359 @@
+		return nil // nothing to serve via S3 gateway
+	}
+	defer func() {
+		panic(
+			fmt.Sprintf("sidecar s3 gateway is exiting; this should never happen (err: %v)", retErr),
+		)
+	}()
+
+	return backoff.RetryNotify(func() (retErr error) {
+		return s.Serve()
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		logrus.Errorf("sidecar s3 gateway: watch closed or error monitoring jobs: %v; retrying in %v", err, d)
+		return nil
+	})
+}
+
+func (s *sidecarS3G) jobInputToBuckets(input *pps.Input) []*s3.Bucket {
+	var buckets []*s3.Bucket
+	// inputToBuckets is a recursive helper
+	pps.VisitInput(input, func(in *pps.Input) {
+		if in.Pfs != nil && in.Pfs.S3 {
+			buckets = append(buckets, &s3.Bucket{
+				Repo:   in.Pfs.Repo,
+				Commit: in.Pfs.Commit,
+				Name:   in.Pfs.Name,
+			})
+		}
+	})
+	return buckets
+}
+
+func (s *sidecarS3G) Serve() error {
+	// Watch for new jobs & initialize s3g for each new job
+	s.HandleJobs(JobHandler{
+		Begin: func(ctx context.Context, jobInfo *pps.JobInfo) error {
+			jobID := jobInfo.Job.ID
+			// Initialize new S3 gateway
+			var outputBucket *s3.Bucket
+			if s.pipelineInfo.S3Out == true {
+				outputBucket = &s3.Bucket{
+					Repo:   jobInfo.OutputCommit.Repo.Name,
+					Commit: jobInfo.OutputCommit.ID,
+					Name:   "out",
+				}
+			}
+			driver := s3.NewWorkerDriver(s.jobInputToBuckets(jobInfo.Input), outputBucket)
+
+			// server new S3 gateway & add to s.servers
+			s.serversMu.Lock()
+			defer s.serversMu.Unlock()
+			// TODO(msteffen) always serve on the same port for now (there shouldn't be
+			// more than one job in s.servers). When parallel jobs are implemented, the
+			// servers in s.servers won't actually serve anymore, and instead parent
+			// server will forward requests based on the request hostname
+			port := s.apiServer.env.S3GatewayPort
+			server, err := s3.Server(port, driver, s)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: couldn't initialize s3 gateway server: %v", err)
+			}
+			strport := strconv.FormatInt(int64(port), 10)
+			listener, err := net.Listen("tcp", ":"+strport)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: could not serve on port %q: %v", port, err)
+			}
+			go func() {
+				server.Serve(listener)
+			}()
+			s.servers[jobID] = server
+			return nil
+		},
+
+		End: func(jobID string) error {
+			s.serversMu.Lock()
+			defer s.serversMu.Unlock()
+			// kill server
+			server, ok := s.servers[jobID]
+			if !ok {
+				// Note that because we call InspectJob after receiving a creation event, we
+				// may never create a service for a job that is created and then immediately
+				// deleted, so no error is returned if jobID is not in s.servers
+				return nil
+			}
+			if err := server.Close(); err != nil {
+				return fmt.Errorf("could not kill sidecar s3 gateway server for job %q: %v", jobID, err)
+			}
+			delete(s.servers, jobID)
```

Comment:
if there is an error closing the server, it'll never be deleted from the map, resulting in a memory leak

Reply:
I updated this section of code. Let me know what you think. At this point, I think the only way that `delete` isn't reached is if the sidecar panics (which it does if it can't `Shutdown` the server correctly)

Reply:
I'm marking this resolved because the code has changed a fair bit (it's now in `s3g_sidecar.go:254`) but rereading I feel confident that if the http server shuts down successfully, then it will be removed from `s.servers` (and if it doesn't shut down successfully then the sidecar will fully restart).